// @deck.gl/carto/dist/es5/layers/quadbin-tileset-2d.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _geoLayers = require("@deck.gl/geo-layers");

var _quadbinUtils = require("./quadbin-utils");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var QuadbinTileset2D = function (_Tileset2D) {
  (0, _inherits2.default)(QuadbinTileset2D, _Tileset2D);

  var _super = _createSuper(QuadbinTileset2D);

  function QuadbinTileset2D() {
    (0, _classCallCheck2.default)(this, QuadbinTileset2D);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(QuadbinTileset2D, [{
    key: "getTileIndices",
    value: function getTileIndices(opts) {
      return (0, _get2.default)((0, _getPrototypeOf2.default)(QuadbinTileset2D.prototype), "getTileIndices", this).call(this, opts).map(_quadbinUtils.tileToQuadbin).map(function (i) {
        return {
          i: i
        };
      });
    }
  }, {
    key: "getTileId",
    value: function getTileId(_ref) {
      var i = _ref.i;
      return i;
    }
  }, {
    key: "getTileMetadata",
    value: function getTileMetadata(_ref2) {
      var i = _ref2.i;
      return (0, _get2.default)((0, _getPrototypeOf2.default)(QuadbinTileset2D.prototype), "getTileMetadata", this).call(this, (0, _quadbinUtils.quadbinToTile)(i));
    }
  }, {
    key: "getTileZoom",
    value: function getTileZoom(_ref3) {
      var i = _ref3.i;
      return Number((0, _quadbinUtils.quadbinZoom)(i));
    }
  }, {
    key: "getParentIndex",
    value: function getParentIndex(_ref4) {
      var i = _ref4.i;
      return {
        i: (0, _quadbinUtils.quadbinParent)(i)
      };
    }
  }]);
  return QuadbinTileset2D;
}(_geoLayers._Tileset2D);

exports.default = QuadbinTileset2D;
