// @loaders.gl/loader-utils/dist/lib/node/fs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._readToArrayBuffer = exports.isSupported = exports.fstat = exports.read = exports.close = exports.open = exports.writeFileSync = exports.writeFile = exports.readFileSync = exports.readFile = exports.stat = exports.readdir = void 0;
// fs wrapper (promisified fs + avoids bundling fs in browsers)
const fs_1 = __importDefault(require("fs"));
const buffer_1 = require("./buffer");
const util_1 = require("./util");
exports.isSupported = Boolean(fs_1.default);
// paths
try {
    /** Wrapper for Node.js fs method */
    exports.readdir = (0, util_1.promisify)(fs_1.default.readdir);
    /** Wrapper for Node.js fs method */
    exports.stat = (0, util_1.promisify)(fs_1.default.stat);
    /** Wrapper for Node.js fs method */
    exports.readFile = (0, util_1.promisify)(fs_1.default.readFile);
    /** Wrapper for Node.js fs method */
    exports.readFileSync = fs_1.default.readFileSync;
    /** Wrapper for Node.js fs method */
    exports.writeFile = (0, util_1.promisify)(fs_1.default.writeFile);
    /** Wrapper for Node.js fs method */
    exports.writeFileSync = fs_1.default.writeFileSync;
    // file descriptors
    /** Wrapper for Node.js fs method */
    exports.open = (0, util_1.promisify)(fs_1.default.open);
    /** Wrapper for Node.js fs method */
    exports.close = (0, util_1.promisify)(fs_1.default.close);
    /** Wrapper for Node.js fs method */
    exports.read = (0, util_1.promisify)(fs_1.default.read);
    /** Wrapper for Node.js fs method */
    exports.fstat = (0, util_1.promisify)(fs_1.default.fstat);
    exports.isSupported = Boolean(fs_1.default);
}
catch {
    // ignore
}
async function _readToArrayBuffer(fd, start, length) {
    const buffer = Buffer.alloc(length);
    const { bytesRead } = await (0, exports.read)(fd, buffer, 0, length, start);
    if (bytesRead !== length) {
        throw new Error('fs.read failed');
    }
    return (0, buffer_1.toArrayBuffer)(buffer);
}
exports._readToArrayBuffer = _readToArrayBuffer;
